devdocs/PIPELINE_REFACTORING_PLAN.md

```diff
@@ -2,276 +2,481 @@
+   - Geometry specification (from Phase 2)
+   - Temporal specification (from Phase 3)
+   - Reader registry (empty for now)
+   - Processing ledger
+2. **Create new `Pipeline` class** that uses state
+3. **Update pipeline operations** to use specifications:
+   - Operations read the specs to know what to do
+   - Specs tell WHERE to get data and WHAT type to create
+   - Operations handle the HOW of extraction and formatting
+4. **Ensure all actions are pure functions** that return new state
+
+**Benefit if stopped here**: Clear separation between decision-making (specs) and execution (operations)
+
+### Phase 5: Reader Interface Definition
+**Standalone Value**: Standardized interface for all metadata extraction
+
+1. **Define new granular reader interfaces**:
+   - `TemporalReader`: Extracts temporal metadata
+   - `SpatialReader`: Extracts spatial metadata
+   - `AttributeReader`: Extracts general metadata/attributes
+2. **Create `ReaderRegistry`** to manage reader selection
+3. **Define reader selection logic** based on file types
+4. **Update pipeline state** to include reader registry
+
+**Benefit if stopped here**: Clear contract for all readers, foundation for extensibility
+
+### Phase 6: Custom Reader Implementation
+**Standalone Value**: Support for any file format without code changes
+
+1. **Implement custom reader base classes**:
+   - `CustomTemporalReader`
+   - `CustomSpatialReader`
+   - `CustomAttributeReader`
+2. **Add configuration parsing** for custom readers
+3. **Implement subprocess execution** with JSON output parsing
+4. **Add timeout and error handling**
+5. **Create example custom readers** and documentation
+6. **Register custom readers** in the reader registry
+
+**Benefit if stopped here**: Users can extend MetGenC to support any file format
+
+### Phase 7: Existing Reader Adaptation
+**Standalone Value**: All readers use consistent interface
+
+1. **Refactor `NetCDFReader`** to implement new interfaces:
+   - Split into temporal, spatial, and attribute readers
+2. **Refactor `CSVReader`** similarly
+3. **Refactor `SnowExReader`** similarly
+4. **Update reader registry** with refactored readers
+5. **Remove old reader code**
+
+**Benefit if stopped here**: Consistent reader architecture throughout the system
+
+### Phase 8: Specify/Execute Separation
+**Standalone Value**: Pure pipeline logic with deferred execution
+
+1. **Define side effect specifications**:
+   - `CreateUMMG`, `CreateCNM`, `WriteFile`, `S3Upload`, `SendMessage`
+2. **Modify pipeline to return specifications** instead of executing
+3. **Create first mapping phase** over granules to collect all specifications
+4. **Create second mapping phase** over specifications to execute with side-effects
+5. **Update ledger** during execution phase
+6. **Implement execution strategies** (parallel, batch, sequential)
+
+**Benefit if stopped here**: Complete separation of pure logic from side effects
+
+### Final Architecture (After Phase 8)
```

Comment:
I would move this up to the section with the current architecture diagram and the new, proposed "conceptual flow."